### PR TITLE
Various refactorings and fixes

### PR DESCRIPTION
**Pull request by daandemeyer**
Split out of #33965

### CODE REVIEW

File: src/core/core-varlink.c [hunk 1]
```diff
@@ -95,6 +95,138 @@ static int build_managed_oom_json_array_element(Unit *u, const char *property, s
...
+        assert(m->managed_oom_varlink);
+
+        r = build_managed_oom_cgroups_json(m, &v);
+        if (r < 0)
+                return r;
+
+        return sd_varlink_send(m->managed_oom_varlink, "io.systemd.oom.ReportManagedOOMCGroups", v);
+}
+
+static int vl_reply(sd_varlink *link, sd_json_variant *parameters, const char *error_id, sd_varlink_reply_flags_t flags, void *userdata) {
```
Comment (yuwata) [inline src/core/core-varlink.c]: Not a matter, but this is for connection to oomd. Maybe better to rename something like `vl_oom_reply()` or `vl_reply_oom()` ?

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/core/core-varlink.c [hunk 2]
```diff
@@ -95,6 +95,138 @@ static int build_managed_oom_json_array_element(Unit *u, const char *property, s
...
+        Manager *m = ASSERT_PTR(userdata);
+        int r;
+
+        if (error_id)
+                log_debug("varlink systemd-oomd client error: %s", error_id);
+
+        if (FLAGS_SET(flags, SD_VARLINK_REPLY_ERROR) && FLAGS_SET(flags, SD_VARLINK_REPLY_LOCAL)) {
```
Comment (YHNdnzj) [inline src/core/core-varlink.c]: If you touch this, consider cherry-picking/applying https://github.com/systemd/systemd/pull/32487/commits/0136871aba7f00630f5fc2e37f5eeae124f6bfa3

File: src/core/core-varlink.c [hunk 3]
```diff
@@ -95,6 +95,138 @@ static int build_managed_oom_json_array_element(Unit *u, const char *property, s
...
+                /* sd_varlink connection was closed, likely because of systemd-oomd restart. Let's try to
+                 * reconnect and send the initial ManagedOOM update again. */
+
+                m->managed_oom_varlink = sd_varlink_unref(link);
+
+                log_debug("Reconnecting to %s", VARLINK_ADDR_PATH_MANAGED_OOM_USER);
+
+                r = manager_varlink_init(m);
```
Comment (YHNdnzj) [inline src/core/core-varlink.c]: The change here seems lost?

File: src/core/unit.c [hunk 4]
```diff
@@ -6473,6 +6470,20 @@ int unit_arm_timer(
...
 
+bool unit_is_filtered(Unit *u, char *const *states, char *const *patterns) {
+        assert(u);
+
+        char **unit_states = STRV_MAKE(
+                        unit_load_state_to_string(u->load_state),
+                        unit_active_state_to_string(unit_active_state(u)),
+                        unit_sub_state_to_string(u));
+
+        if (!strv_isempty(states) && !strv_overlap(states, unit_states))
```
Comment (YHNdnzj) [inline src/core/unit.c]: If `states` is empty, there's no need to acquire unit states. Hence move the STRV_MAKE into the if branch?

File: src/core/unit.c [hunk 5]
```diff
@@ -6473,6 +6470,20 @@ int unit_arm_timer(
         return 0;
 }
 
+bool unit_is_filtered(Unit *u, char *const *states, char *const *patterns) {
+        assert(u);
+
+        char **unit_states = STRV_MAKE(
```
Comment (YHNdnzj) [inline src/core/unit.c]:
```suggestion
        char * const *unit_states = STRV_MAKE(
```

**Review by YHNdnzj — APPROVED** [review]

**Pull request merged**